app/src/main/java/android/support/v17/leanback/supportleanbackshowcase/app/media/ExoPlayerAdapter.java

```diff
@@ -0,0 +1,312 @@
+                mBufferingStart || !mInitialized);
+    }
+
+    /**
+     * Constructor.
+     */
+    public ExoPlayerAdapter(Context context) {
+        mContext = context;
+        mPlayer = ExoPlayerFactory.newSimpleInstance(mContext,
+                new DefaultTrackSelector(),
+                new DefaultLoadControl());
+        mPlayer.addListener(this);
+    }
+
+    @Override
+    public void onAttachedToHost(PlaybackGlueHost host) {
+        if (host instanceof SurfaceHolderGlueHost) {
+            mSurfaceHolderGlueHost = ((SurfaceHolderGlueHost) host);
+            mSurfaceHolderGlueHost.setSurfaceHolderCallback(new VideoPlayerSurfaceHolderCallback());
+        }
+    }
+
+    /**
+     * Will reset the {@link ExoPlayer} and the glue such that a new file can be played. You are
+     * not required to call this method before playing the first file. However you have to call it
+     * before playing a second one.
+     */
+    public void reset() {
+        changeToUninitialized();
+        mPlayer.stop();
+    }
+
+    void changeToUninitialized() {
+        if (mInitialized) {
+            mInitialized = false;
+            notifyBufferingStartEnd();
+            if (mHasDisplay) {
+                getCallback().onPreparedStateChanged(ExoPlayerAdapter.this);
+            }
+        }
+    }
+
+    /**
+     * Release internal MediaPlayer. Should not use the object after call release().
+     */
+    public void release() {
+        changeToUninitialized();
+        mHasDisplay = false;
+        mPlayer.release();
+    }
+
+    @Override
+    public void onDetachedFromHost() {
+        if (mSurfaceHolderGlueHost != null) {
+            mSurfaceHolderGlueHost.setSurfaceHolderCallback(null);
+            mSurfaceHolderGlueHost = null;
+        }
+        reset();
+        release();
+    }
+
+    /**
+     * @see SimpleExoPlayer#setVideoSurfaceHolder(SurfaceHolder)
+     */
+    void setDisplay(SurfaceHolder surfaceHolder) {
+        boolean hadDisplay = mHasDisplay;
+        mHasDisplay = surfaceHolder != null;
+        if (hadDisplay == mHasDisplay) {
+            return;
+        }
+
+        mPlayer.setVideoSurfaceHolder(surfaceHolder);
+        if (mHasDisplay) {
+            if (mInitialized) {
+                getCallback().onPreparedStateChanged(ExoPlayerAdapter.this);
+            }
+        } else {
+            if (mInitialized) {
+                getCallback().onPreparedStateChanged(ExoPlayerAdapter.this);
+            }
+        }
+    }
+
+    @Override
+    public void setProgressUpdatingEnabled(final boolean enabled) {
+        mHandler.removeCallbacks(mRunnable);
+        if (!enabled) {
+            return;
+        }
+        mHandler.postDelayed(mRunnable, getUpdatePeriod());
+    }
+
+    int getUpdatePeriod() {
+        return 16;
+    }
+
+    @Override
+    public boolean isPlaying() {
+        boolean exoPlayerIsPlaying = mPlayer.getPlaybackState() == ExoPlayer.STATE_READY
```

Comment:
when it's buffering,  we still consider it as playing state (and UI button wont flip the icon)